# Pandas：从公司财务表到可分析数据

Pandas 是 Python 中处理表格数据的核心工具。经管数据常见的形态是一行一条记录、一列一个变量，例如上市公司年度数据、问卷数据、交易数据和地区统计数据。把这些数据读成 pandas 的 `DataFrame` 后，可以继续完成查看、筛选、变量生成、清洗、合并、分组统计、重塑和时间顺序计算。

学习 pandas 时，函数本身通常不难，关键是每一步都要确认数据是否符合预期。读入后看一眼，筛选后看一眼，生成新列后再看一眼。这样做可以及时发现列选错、类型不对、缺失值异常、重复记录等问题。


**Jupyter Notebook 简介**

本章使用 Jupyter Notebook。Notebook 文件扩展名是 `.ipynb`，适合记录数据分析过程：代码、文字说明和运行结果可以放在同一个文件中。

**数据存放约定**

本课程约定：数据文件都放在工作目录下的 `data` 文件夹中。工作目录可以理解为当前项目或当前 notebook 所在的主要目录。读取数据时，可以用相对路径引用文件，例如 `data/company_annual_operations_clean.xlsx`。

**本章使用的数据**

本章主要使用两张教学表：

- `data/company_annual_operations_clean.xlsx`：干净的公司年度经营数据。一行是一家公司某一年的记录，包含公司名称、日期、行业、地区、资产、收入、利润、员工数等变量，用于练习单表操作。
- `data/financial_indicators_dirty.xlsx`：待清洗的公司年度财务指标数据。它和第一张表同样以公司和年份为核心，但故意保留了一些常见问题，用于练习清洗，并在清洗后与经营表合并。


## 本章知识点安排

**阶段 1：数据读取与初步查看**

读入公司年度经营表，查看前几行、后几行、行列数、列名、数据类型和描述统计。先知道表里有什么，再决定后面怎么处理。

**阶段 2：行列选择与条件筛选**

围绕“在行与列上操作”展开：选单列、选多列、`iloc`、`loc`、条件筛选、复合条件、同时选行和列、`query()`，以及筛选后查看结果。

**阶段 3：变量生成与数据修改**

用已有列生成新列，做列运算，覆盖已有列，用日期列生成年度和上市年限，用 `np.where()`、`pd.cut()` 和带筛选的 `.loc` 赋值，排序、排名，并做简单统计。

**阶段 4：常见数据问题处理**

读入待清洗的财务指标表，检查缺失值和重复值，处理证券代码、日期、特殊文本和数值类型，使用 `dropna()`、`fillna()`，并理解链式赋值 warning。

**阶段 5：表的追加与合并**

使用 `concat()` 追加同结构表，用 `merge()` 按连接键合并经营表和财务指标表，并检查合并结果。

**阶段 6：分组汇总与组内比较**

从公司明细表转向分组结论：`groupby()`、多指标 `agg()`、自定义聚合函数、`transform()`、每组取前几名和分组循环。

**阶段 7：数据重塑与结果表构造**

理解长表和宽表，使用 `pivot_table()`、`set_index()`、`reset_index()`，构造 `Series` / `DataFrame`，并导出结果。

**阶段 8：公司年度面板中的时间顺序**

让 pandas 识别公司内部的时间顺序，使用排序、日期索引、日期切片、`shift()`、`diff()`、`pct_change()` 和 `cumprod()` 观察年度变化。

**开始前：DataFrame 和 Series**

Pandas 主要处理二维表格。一个 `DataFrame` 可以理解为一张 Excel 表：有行、列、列名和行索引。一个 `Series` 可以理解为一列数据，它由索引和值组成。多个 `Series` 横向放在一起，就形成一个 `DataFrame`。

```text
index + ndarray -> Series
index + Series + Series + ... -> DataFrame
```

![](images/df-dp.png)

**Excel 和 CSV**

常见表格数据主要有两类：

- Excel 文件：扩展名通常是 `.xlsx`。可以保存格式、颜色、多个工作表等信息，适合给人查看和编辑。
- CSV 文件：扩展名是 `.csv`。本质上是纯文本，只保存数据本身，不保存格式。它体积小、通用性强，几乎所有数据软件都能读取。

在数据分析中，Excel 和 CSV 都常见。拿不准保存成什么格式时，CSV 往往更通用；需要给人直接打开查看时，Excel 更方便。

## 阶段 1：数据读取与初步查看

拿到一张新表时，先不要急着计算。第一步是确认数据是否读对了：表有多大、有哪些列、每列大致是什么类型、数值范围是否看起来合理。本阶段先读入公司年度经营表，并建立对这张表的整体认识。

**读取数据，并看表的样子**

1. `pd.read_excel("文件路径")`：读取 Excel 文件，得到一张 `DataFrame`。

2. `head()` / `tail()`：分别查看开头和末尾，默认显示 5 行；`head(10)` / `tail(10)` 可以指定行数。

In [1]:
import pandas as pd  # 表格数据处理
import numpy as np  # 数值和缺失值处理

pd.set_option("display.max_columns", 30)  # 最多显示 30 列
pd.set_option("display.float_format", "{:.2f}".format)  # 小数显示为 2 位

annual_raw = pd.read_excel("data/company_annual_operations_clean.xlsx")  # 读取 Excel
annual_raw.head()  # 查看前 5 行

,证券代码,股票简称,统计截止日期,年报公布日期,行业名称,所属省份,所属城市,首次上市日期,上市状态,总资产_亿元,总负债_亿元,营业收入_亿元,营业成本_亿元,销售费用_亿元,净利润_亿元,经营现金流_亿元,资产负债率,营业毛利率,员工数目
0,2,万科A,2018-12-31,2019-03-26,房地产业,广东省,深圳市,1991-01-29,正常上市,15285.79,12929.59,2976.79,1861.04,78.68,492.72,336.18,0.85,0.37,104300
1,2,万科A,2019-12-31,2020-03-18,房地产业,广东省,深圳市,1991-01-29,正常上市,17299.29,14593.50,3678.94,2345.50,90.44,551.32,456.87,0.84,0.36,131505
2,2,万科A,2020-12-31,2021-03-31,房地产业,广东省,深圳市,1991-01-29,正常上市,18691.77,15193.33,4191.12,2965.41,106.37,592.98,531.88,0.81,0.29,140565
3,4,国农科技,2018-12-31,2019-04-26,医药制造业,广东省,深圳市,1991-01-14,正常上市,3.51,1.68,3.67,0.66,2.71,-0.22,-0.53,0.48,0.82,210
4,4,国农科技,2019-12-31,2020-04-28,软件和信息技术服务业,广东省,深圳市,1991-01-14,正常上市,14.94,0.90,1.08,0.29,0.83,-0.04,-0.16,0.06,0.73,251


In [2]:
annual_raw.tail()  # 查看后 5 行

,证券代码,股票简称,统计截止日期,年报公布日期,行业名称,所属省份,所属城市,首次上市日期,上市状态,总资产_亿元,总负债_亿元,营业收入_亿元,营业成本_亿元,销售费用_亿元,净利润_亿元,经营现金流_亿元,资产负债率,营业毛利率,员工数目
64,638,万方发展,2019-12-31,2020-06-29,软件和信息技术服务业,吉林省,白山市,1996-11-26,正常上市,5.28,3.03,1.16,0.62,0.11,0.16,0.04,0.57,0.46,466
65,638,万方发展,2020-12-31,2021-04-14,软件和信息技术服务业,吉林省,白山市,1996-11-26,正常上市,5.07,3.39,1.11,0.80,0.13,-0.20,-0.02,0.67,0.28,670
66,682,东方电子,2018-12-31,2019-04-02,电气机械及器材制造业,山东省,烟台市,1997-01-21,正常上市,51.55,18.73,30.42,20.58,3.58,2.23,2.66,0.36,0.32,4564
67,682,东方电子,2019-12-31,2020-04-15,软件和信息技术服务业,山东省,烟台市,1997-01-21,正常上市,62.61,27.04,34.19,22.27,4.30,2.83,5.09,0.43,0.35,5087
68,682,东方电子,2020-12-31,2021-04-23,软件和信息技术服务业,山东省,烟台市,1997-01-21,正常上市,68.98,30.97,37.19,24.40,4.52,3.17,3.20,0.45,0.34,5665


**查看表的结构和数值概况**

1. `shape`：查看行数和列数。

2. `columns.tolist()`：查看列名，并转换成普通列表，显示起来更直接。

3. `dtypes` / `info()`：查看数据类型；`info()` 还会显示每列的非缺失值数量。

4. `describe()`：查看数值列的样本数、均值、标准差、最小值、四分位数和最大值。编号和日期通常不放进描述统计。

In [3]:
annual_raw.shape  # 查看行数和列数

(69, 19)

In [4]:
annual_raw.columns.tolist()  # 查看列名列表

['证券代码',
 '股票简称',
 '统计截止日期',
 '年报公布日期',
 '行业名称',
 '所属省份',
 '所属城市',
 '首次上市日期',
 '上市状态',
 '总资产_亿元',
 '总负债_亿元',
 '营业收入_亿元',
 '营业成本_亿元',
 '销售费用_亿元',
 '净利润_亿元',
 '经营现金流_亿元',
 '资产负债率',
 '营业毛利率',
 '员工数目']

In [5]:
annual_raw.dtypes  # 查看每列数据类型

证券代码                 int64
股票简称                object
统计截止日期      datetime64[ns]
年报公布日期      datetime64[ns]
行业名称                object
所属省份                object
所属城市                object
首次上市日期      datetime64[ns]
上市状态                object
总资产_亿元             float64
总负债_亿元             float64
营业收入_亿元            float64
营业成本_亿元            float64
销售费用_亿元            float64
净利润_亿元             float64
经营现金流_亿元           float64
资产负债率              float64
营业毛利率              float64
员工数目                 int64
dtype: object

In [6]:
annual_raw.info()  # 查看表结构摘要

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69 entries, 0 to 68
Data columns (total 19 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   证券代码      69 non-null     int64         
 1   股票简称      69 non-null     object        
 2   统计截止日期    69 non-null     datetime64[ns]
 3   年报公布日期    69 non-null     datetime64[ns]
 4   行业名称      69 non-null     object        
 5   所属省份      69 non-null     object        
 6   所属城市      69 non-null     object        
 7   首次上市日期    69 non-null     datetime64[ns]
 8   上市状态      69 non-null     object        
 9   总资产_亿元    69 non-null     float64       
 10  总负债_亿元    69 non-null     float64       
 11  营业收入_亿元   69 non-null     float64       
 12  营业成本_亿元   69 non-null     float64       
 13  销售费用_亿元   69 non-null     float64       
 14  净利润_亿元    69 non-null     float64       
 15  经营现金流_亿元  69 non-null     float64       
 16  资产负债率     69 non-null     float64       
 17  营业毛利率     69 non-n

In [7]:
business_cols = [
    "总资产_亿元",
    "总负债_亿元",
    "营业收入_亿元",
    "营业成本_亿元",
    "销售费用_亿元",
    "净利润_亿元",
    "经营现金流_亿元",
    "资产负债率",
    "营业毛利率",
    "员工数目",
]  # 只保留适合做描述统计的经营数值列

annual_raw[business_cols].describe()  # 查看经营数值列描述统计

,总资产_亿元,总负债_亿元,营业收入_亿元,营业成本_亿元,销售费用_亿元,净利润_亿元,经营现金流_亿元,资产负债率,营业毛利率,员工数目
count,69.00,69.00,69.00,69.00,69.00,69.00,69.00,69.00,69.00,69.00
mean,1072.96,830.10,432.11,318.19,25.86,42.71,45.99,0.53,0.29,18871.78
std,3536.24,2932.35,943.09,662.07,67.57,121.33,117.44,0.18,0.19,35995.12
min,3.51,0.90,1.08,0.29,0.05,-4.45,-32.30,0.06,0.05,148.00
25%,36.36,11.73,21.58,13.87,0.52,0.61,0.04,0.43,0.14,1132.00
50%,109.50,59.09,57.06,38.41,2.56,3.82,2.66,0.55,0.22,5504.00
75%,168.06,91.53,172.49,158.35,8.59,8.03,11.24,0.67,0.39,13432.00
max,18691.77,15193.33,4191.12,2965.41,346.11,592.98,531.88,0.85,0.82,149239.00


**建立工作副本**

拿到原始数据后，建议复制一张工作副本。后面的筛选、删除、修改都在副本上做；如果处理过程中改错了，还可以回到原始表重新开始。

1. `drop(columns=[...])`：删除指定列。很多 pandas 操作会返回一张新表，原表通常不会自动改变。

2. `.copy()`：复制出工作副本。这里先暂时放下 `证券代码`，集中观察公司经营变量。

In [8]:
annual = annual_raw.drop(columns=["证券代码"]).copy()  # 删除暂不用的列并复制副本
annual.head()  # 查看工作副本

,股票简称,统计截止日期,年报公布日期,行业名称,所属省份,所属城市,首次上市日期,上市状态,总资产_亿元,总负债_亿元,营业收入_亿元,营业成本_亿元,销售费用_亿元,净利润_亿元,经营现金流_亿元,资产负债率,营业毛利率,员工数目
0,万科A,2018-12-31,2019-03-26,房地产业,广东省,深圳市,1991-01-29,正常上市,15285.79,12929.59,2976.79,1861.04,78.68,492.72,336.18,0.85,0.37,104300
1,万科A,2019-12-31,2020-03-18,房地产业,广东省,深圳市,1991-01-29,正常上市,17299.29,14593.50,3678.94,2345.50,90.44,551.32,456.87,0.84,0.36,131505
2,万科A,2020-12-31,2021-03-31,房地产业,广东省,深圳市,1991-01-29,正常上市,18691.77,15193.33,4191.12,2965.41,106.37,592.98,531.88,0.81,0.29,140565
3,国农科技,2018-12-31,2019-04-26,医药制造业,广东省,深圳市,1991-01-14,正常上市,3.51,1.68,3.67,0.66,2.71,-0.22,-0.53,0.48,0.82,210
4,国农科技,2019-12-31,2020-04-28,软件和信息技术服务业,广东省,深圳市,1991-01-14,正常上市,14.94,0.90,1.08,0.29,0.83,-0.04,-0.16,0.06,0.73,251


阶段 1 小结：拿到表格后，先用 `head()`、`tail()`、`shape`、`columns`、`dtypes`、`info()` 和 `describe()` 看结构、类型和大致分布。现在我们已经知道这张表的行列、变量和主要数值范围，下一步就可以从表里切出真正要看的部分。

**练习**：完成最后的阶段 1 练习。

## 阶段 2：行列选择与条件筛选

多数情况下，我们要处理的数据只是现有数据的一部分，所以需要先按分析要求把数据切出来。比如在这张公司年度经营表中，如果想观察 2020 年制造业中盈利、收入数据完整、负债率相对不高的公司，就要先筛选出这些公司，再只保留后面要比较的经营列。这里练的是一件事：从一张完整表里取出“现在需要”的行和列。

范例：选出 2020 年、行业名称包含“制造”、净利润为正、营业收入不缺失、资产负债率低于 0.7 的公司年度记录，作为下一步观察的名单。

**DataFrame 和 Series 的结构**

在 pandas 中，一张表可以拆成几个部分来看：

1. `df.index`：行索引，用来标识每一行。

2. `df.columns`：列名，用来标识每一列。

3. `df.values`：表格中的值。

4. `df["A"]`：从 DataFrame 中取出一列，会得到一个 `Series`。`Series` 也有自己的 `index` 和 `values`。

![](images/df_structure_breakdown.png)

**选择列**

1. `df["列名"]`：用列名选择单列，返回 `Series`。`Series` 可以理解为一列数据。

2. `df[[列名列表]]`：用列名的 list 选择多列，返回 `DataFrame`。这里可以用上前面学过的 list 写法，例如 `["股票简称", "行业名称"]`。

3. `type()`：查看对象类型。这里用它区分 `Series` 和 `DataFrame`。

In [9]:
one_col = annual["营业收入_亿元"]  # 用列名选择单列
one_col.head()  # 查看单列数据

0   2976.79
1   3678.94
2   4191.12
3      3.67
4      1.08
Name: 营业收入_亿元, dtype: float64

In [10]:
type(one_col)  # 查看单列对象类型

pandas.core.series.Series

In [11]:
some_cols = annual[["股票简称", "行业名称", "营业收入_亿元"]]  # 用列名 list 选择多列
some_cols.head()  # 查看多列数据

,股票简称,行业名称,营业收入_亿元
0,万科A,房地产业,2976.79
1,万科A,房地产业,3678.94
2,万科A,房地产业,4191.12
3,国农科技,医药制造业,3.67
4,国农科技,软件和信息技术服务业,1.08


In [12]:
type(some_cols)  # 查看多列对象类型

pandas.core.frame.DataFrame

In [13]:
annual[["股票简称", "统计截止日期", "行业名称", "营业收入_亿元", "净利润_亿元"]].head()  # 查看多列结果

,股票简称,统计截止日期,行业名称,营业收入_亿元,净利润_亿元
0,万科A,2018-12-31,房地产业,2976.79,492.72
1,万科A,2019-12-31,房地产业,3678.94,551.32
2,万科A,2020-12-31,房地产业,4191.12,592.98
3,国农科技,2018-12-31,医药制造业,3.67,-0.22
4,国农科技,2019-12-31,软件和信息技术服务业,1.08,-0.04


**按位置和标签选择**

1. `.iloc[行位置, 列位置]`：按整数位置选择。行和列都使用从 0 开始的顺序位置。

2. `.iloc` 切片：写法和列表切片类似，`起点:终点` 表示从起点取到终点之前，不包括终点。例如 `:5` 表示位置 0 到 4。

3. `.loc[行标签, 列标签]`：按标签选择。行标签可以是索引标签，也可以是筛选条件；列标签可以是一个列名，也可以是列名列表。

4. `.loc` 标签切片：按标签切片时包含结束点，这和 Python 列表切片不同。

如果行索引看起来刚好也是 0、1、2、3，`.loc[0:3]` 和 `.iloc[:4]` 的结果可能看起来一样。但它们的含义不同：`.loc` 看标签，`.iloc` 看位置。索引被筛选、排序或改成公司代码之后，这个差别会变得很明显。

In [14]:
annual.iloc[:5, :4]  # 前 5 行、前 4 列

,股票简称,统计截止日期,年报公布日期,行业名称
0,万科A,2018-12-31,2019-03-26,房地产业
1,万科A,2019-12-31,2020-03-18,房地产业
2,万科A,2020-12-31,2021-03-31,房地产业
3,国农科技,2018-12-31,2019-04-26,医药制造业
4,国农科技,2019-12-31,2020-04-28,软件和信息技术服务业


In [15]:
# 查看 2020 年公司的几个关键经营列。
annual.loc[
    annual["统计截止日期"].dt.year == 2020,  # 行位置：筛选 2020 年
    ["股票简称", "行业名称", "营业收入_亿元", "净利润_亿元", "资产负债率"],  # 列位置：保留指定列
].head()  # 查看前 5 行

,股票简称,行业名称,营业收入_亿元,净利润_亿元,资产负债率
2,万科A,房地产业,4191.12,592.98,0.81
5,国华网安,软件和信息技术服务业,2.81,0.62,0.07
8,深振业A,房地产业,29.35,9.03,0.49
11,深物业A,房地产业,41.04,7.31,0.69
14,沙河股份,房地产业,3.47,0.09,0.61


In [16]:
annual.loc[0:3, "股票简称":"行业名称"]  # 标签切片包含结束点

,股票简称,统计截止日期,年报公布日期,行业名称
0,万科A,2018-12-31,2019-03-26,房地产业
1,万科A,2019-12-31,2020-03-18,房地产业
2,万科A,2020-12-31,2021-03-31,房地产业
3,国农科技,2018-12-31,2019-04-26,医药制造业


**条件筛选**

1. `df["列"] == 值`：生成一组 `True` / `False`，可以放到 `.loc` 的行位置，保留满足条件的行。

2. `&` / `|` / `~`：分别表示“并且”“或者”“取反”。复合条件中，每个条件外面都要加括号。忘记括号时，pandas 的报错通常不直观，先回头检查每个条件是不是都包在括号里。

3. `.notna()`：判断是否不是缺失值。筛选时常用来保留关键列有值的行。

4. `.str.contains("关键词", na=False)`：判断文本列是否包含关键词；`na=False` 表示缺失值按“不包含”处理。

5. `.dt.year`：从日期列中取出年份。日期筛选中经常用到。

6. `len()`：查看结果有多少行。筛选后通常要看一下行数和前后几行。

In [17]:
# 筛选出 2020 年、制造业、盈利、收入不缺失且负债率较低的记录。
good_2020 = annual.loc[
    (annual["统计截止日期"].dt.year == 2020)  # 条件 1：2020 年
    & (annual["行业名称"].str.contains("制造", na=False))  # 条件 2：行业名称包含“制造”
    & (annual["净利润_亿元"] > 0)  # 条件 3：净利润为正
    & (annual["营业收入_亿元"].notna())  # 条件 4：营业收入不是缺失值
    & (annual["资产负债率"] < 0.7),  # 条件 5：资产负债率小于 0.7
    ["股票简称", "行业名称", "所属省份", "营业收入_亿元", "净利润_亿元", "资产负债率"],  # 保留指定列
]

good_2020.head()  # 查看筛选结果

,股票简称,行业名称,所属省份,营业收入_亿元,净利润_亿元,资产负债率
20,深华发A,计算机、通信和其他电子设备制造业,广东省,6.92,0.07,0.46
23,深科技,计算机、通信和其他电子设备制造业,广东省,149.67,9.53,0.63
26,富奥股份,汽车制造业,吉林省,111.13,9.08,0.43
29,深纺织A,计算机、通信和其他电子设备制造业,广东省,21.09,0.43,0.22
32,德赛电池,电气机械及器材制造业,广东省,193.98,7.40,0.69


In [18]:
len(good_2020)  # 查看筛选结果行数

12

In [19]:
good_2020.tail()  # 查看筛选结果末尾

,股票简称,行业名称,所属省份,营业收入_亿元,净利润_亿元,资产负债率
47,许继电气,电气机械及器材制造业,河南省,111.91,8.03,0.45
50,派林生物,医药制造业,山西省,10.50,1.83,0.41
53,东阿阿胶,医药制造业,山东省,34.09,0.41,0.11
56,江铃汽车,汽车制造业,江西省,330.96,5.51,0.61
62,万向钱潮,汽车制造业,浙江省,108.82,4.44,0.43


**用字符串写筛选条件**

1. `query("条件")`：把筛选条件写成字符串，列名可以直接出现在字符串里。条件较短时，这种写法很方便。

2. `and` / `or` / `not`：在 `query()` 字符串里可以表示“并且”“或者”“取反”。

In [20]:
annual.query("净利润_亿元 > 0 and 资产负债率 < 0.7").head()  # 用 query 筛选

,股票简称,统计截止日期,年报公布日期,行业名称,所属省份,所属城市,首次上市日期,上市状态,总资产_亿元,总负债_亿元,营业收入_亿元,营业成本_亿元,销售费用_亿元,净利润_亿元,经营现金流_亿元,资产负债率,营业毛利率,员工数目
5,国华网安,2020-12-31,2021-04-28,软件和信息技术服务业,广东省,深圳市,1991-01-14,正常上市,15.64,1.07,2.81,0.69,0.51,0.62,0.04,0.07,0.75,264
6,深振业A,2018-12-31,2019-03-30,房地产业,广东省,深圳市,1992-04-27,正常上市,135.37,71.51,25.12,15.38,0.43,9.23,16.07,0.53,0.39,377
7,深振业A,2019-12-31,2020-04-09,房地产业,广东省,深圳市,1992-04-27,正常上市,157.45,85.75,37.31,20.44,0.52,8.43,-1.94,0.54,0.45,385
8,深振业A,2020-12-31,2021-03-31,房地产业,广东省,深圳市,1992-04-27,正常上市,154.35,75.97,29.35,15.67,0.45,9.03,-1.68,0.49,0.47,397
9,深物业A,2018-12-31,2019-03-30,房地产业,广东省,深圳市,1992-03-30,正常上市,58.20,24.79,27.87,13.15,0.47,5.93,11.24,0.43,0.53,5504


阶段 2 小结：想从表中取出一部分数据，可以按列名选列，用 `iloc` 按位置选，用 `loc` 按标签或条件选；条件较短时，`query()` 也很方便。筛选之后用 `head()`、`tail()` 和行数看一眼结果。`good_2020` 是一份观察名单，后面如果要继续分析，就可以围绕这样的名单生成更多变量。

**练习**：完成最后的阶段 2 练习。

## 阶段 3：变量生成与数据修改

选出需要的数据后，很多分析还要把原始字段转换成更接近问题的变量。比如收入、成本和利润可以说明经营规模和结果，但要比较盈利能力，还需要计算净利率；要比较人员效率，可以计算人均营业收入；要观察上市时间，也可以从日期中提取年份并计算上市年限。新增变量后要重点看被改过的列，不需要每次都展示全部列。

范例：在原始公司年度经营数据基础上生成年份、净利率、人均营业收入、上市年限、资产规模、负债水平等新变量，并整理出 2020 年公司经营表现排名。

**用已有列生成新列**

1. `df["新列"] = 表达式`：新增一列，写法和给字典增加新键类似。

2. `列运算`：表达式可以来自一列，也可以来自多列共同计算。例如净利率可以由净利润除以营业收入得到。

3. `日期属性`：日期列可以用 `.dt.year` 取出年份，再继续计算。

4. `比较运算`：例如 `df["净利润_亿元"] > 0` 会得到一列 `True` / `False`，可以直接作为新列。

5. `pd.cut()`：把连续数值分成几个区间，常用于生成“高、中、低”这类分组变量。

列运算完成后，先看几列核心输入和新列是否对得上。比如净利率应当和净利润、营业收入方向一致；上市年限不应该是负数。

In [21]:
# 先从日期列生成年度变量。
annual["年份"] = annual["统计截止日期"].dt.year  # 从日期提取年份
annual["上市年份"] = annual["首次上市日期"].dt.year  # 从日期提取上市年份
annual["上市年限"] = annual["年份"] - annual["上市年份"]  # 计算上市年限

annual[["股票简称", "统计截止日期", "年份", "首次上市日期", "上市年份", "上市年限"]].head()  # 查看日期变量

,股票简称,统计截止日期,年份,首次上市日期,上市年份,上市年限
0,万科A,2018-12-31,2018,1991-01-29,1991,27
1,万科A,2019-12-31,2019,1991-01-29,1991,28
2,万科A,2020-12-31,2020,1991-01-29,1991,29
3,国农科技,2018-12-31,2018,1991-01-14,1991,27
4,国农科技,2019-12-31,2019,1991-01-14,1991,28


In [22]:
# 再根据经营数据生成比较指标和分组标签。
annual["净利率"] = annual["净利润_亿元"] / annual["营业收入_亿元"]  # 用两列计算新列
annual["人均营业收入_万元"] = annual["营业收入_亿元"] * 10000 / annual["员工数目"]  # 用两列计算人均指标
annual["是否盈利"] = annual["净利润_亿元"] > 0  # 比较结果生成布尔列
annual["资产规模"] = pd.cut(
    annual["总资产_亿元"],  # 要分箱的连续变量
    bins=[0, 100, 1000, np.inf],  # 分箱边界
    labels=["小", "中", "大"],  # 每个区间的标签
)

annual[["股票简称", "年份", "营业收入_亿元", "净利润_亿元", "净利率", "人均营业收入_万元", "资产规模"]].head()  # 查看经营变量

,股票简称,年份,营业收入_亿元,净利润_亿元,净利率,人均营业收入_万元,资产规模
0,万科A,2018,2976.79,492.72,0.17,285.41,大
1,万科A,2019,3678.94,551.32,0.15,279.76,大
2,万科A,2020,4191.12,592.98,0.14,298.16,大
3,国农科技,2018,3.67,-0.22,-0.06,174.76,小
4,国农科技,2019,1.08,-0.04,-0.04,43.03,小


**按条件修改数据**

1. `np.where(条件, 条件成立时的值, 条件不成立时的值)`：适合二选一地生成新列。

2. `.loc[条件, 列名] = 新值`：适合对满足条件的行局部修改。

3. `df[条件]["列"] = 新值`：这类连续使用 `[]` 的写法称为链式赋值。pandas 不一定能判断你想修改原表还是临时结果，因此容易触发 warning；有时 warning 后原表并没有被修改。看到这类 warning 时，先把写法改成 `.loc`，再重新检查被修改的列。

4. `df.loc[条件, "列"] = 新值`：需要修改原表时，课堂和作业中建议使用这种写法。

In [23]:
# 根据资产负债率生成负债水平标签。
annual["负债水平"] = np.where(
    annual["资产负债率"] >= 0.7,  # 条件
    "较高",  # 条件成立时的值
    "正常",  # 条件不成立时的值
)

annual[["股票简称", "年份", "资产负债率", "负债水平"]].head()  # 查看条件生成结果

,股票简称,年份,资产负债率,负债水平
0,万科A,2018,0.85,较高
1,万科A,2019,0.84,较高
2,万科A,2020,0.81,较高
3,国农科技,2018,0.48,正常
4,国农科技,2019,0.06,正常


In [24]:
# 用 .loc 把高负债且亏损的记录标记为重点关注。
annual["重点关注"] = False  # 先给整列一个默认值
annual.loc[
    (annual["资产负债率"] >= 0.7) & (annual["净利润_亿元"] < 0),  # 高负债且亏损
    "重点关注",
] = True  # 按条件局部修改
annual[["股票简称", "年份", "资产负债率", "净利润_亿元", "重点关注"]].head()  # 查看局部修改结果

,股票简称,年份,资产负债率,净利润_亿元,重点关注
0,万科A,2018,0.85,492.72,False
1,万科A,2019,0.84,551.32,False
2,万科A,2020,0.81,592.98,False
3,国农科技,2018,0.48,-0.22,False
4,国农科技,2019,0.06,-0.04,False


**排序和排名**

1. `sort_values()`：按一列或多列排序。`ascending=False` 表示从大到小排序；多列排序时可以给 `ascending` 传入列表。

2. `rank()`：生成排名。排序适合查看结果；如果需要把排名作为一列继续使用，可以用 `rank()`。

3. `.astype(int)`：把结果转换为整数类型。排名本身常常希望显示为整数。

In [25]:
# 得到 2020 年公司经营表现排名表。
annual_2020_rank = (
    annual.loc[annual["年份"] == 2020].copy()  # 筛选 2020 年并复制
    .sort_values(["营业收入_亿元", "净利率"], ascending=[False, False])  # 按收入和净利率降序
)

annual_2020_rank[[
    "股票简称", "行业名称", "营业收入_亿元", "净利润_亿元", "净利率", "资产负债率", "负债水平"
]].head(10)  # 查看排名靠前的公司

,股票简称,行业名称,营业收入_亿元,净利润_亿元,净利率,资产负债率,负债水平
2,万科A,房地产业,4191.12,592.98,0.14,0.81,较高
41,美的集团,电气机械及器材制造业,2842.21,275.07,0.10,0.66,正常
44,潍柴动力,汽车制造业,1974.91,112.75,0.06,0.70,较高
17,深康佳A,计算机、通信和其他电子设备制造业,503.52,5.40,0.01,0.79,较高
56,江铃汽车,汽车制造业,330.96,5.51,0.02,0.61,正常
32,德赛电池,电气机械及器材制造业,193.98,7.40,0.04,0.69,正常
23,深科技,计算机、通信和其他电子设备制造业,149.67,9.53,0.06,0.63,正常
47,许继电气,电气机械及器材制造业,111.91,8.03,0.07,0.45,正常
26,富奥股份,汽车制造业,111.13,9.08,0.08,0.43,正常
62,万向钱潮,汽车制造业,108.82,4.44,0.04,0.43,正常


In [26]:
# 把营业收入排名保存成一列，便于后续继续使用。
annual_2020_rank["收入排名"] = (
    annual_2020_rank["营业收入_亿元"]
    .rank(ascending=False, method="min")  # 按营业收入从大到小排名
    .astype(int)  # 转成整数
)

annual_2020_rank[["股票简称", "营业收入_亿元", "收入排名"]].head(10)  # 查看排名列

,股票简称,营业收入_亿元,收入排名
2,万科A,4191.12,1
41,美的集团,2842.21,2
44,潍柴动力,1974.91,3
17,深康佳A,503.52,4
56,江铃汽车,330.96,5
32,德赛电池,193.98,6
23,深科技,149.67,7
47,许继电气,111.91,8
26,富奥股份,111.13,9
62,万向钱潮,108.82,10


In [27]:
# 查看排名尾部，理解结果的另一端。
annual_2020_rank[[
    "股票简称", "营业收入_亿元", "净利润_亿元", "净利率", "资产负债率"
]].tail(5)  # 查看排名尾部

,股票简称,营业收入_亿元,净利润_亿元,净利率,资产负债率
50,派林生物,10.50,1.83,0.17,0.41
20,深华发A,6.92,0.07,0.01,0.46
14,沙河股份,3.47,0.09,0.03,0.61
5,国华网安,2.81,0.62,0.22,0.07
65,万方发展,1.11,-0.20,-0.18,0.67


**简单统计和展示列名**

1. `nunique()`：计算不重复取值数量。这里用于计算公司数量。

2. `mean()` / `max()`：分别计算平均值和最大值。

3. `describe()`：查看多列描述统计。

4. `rename(columns={...})`：重命名列。常用于把结果表整理成更适合展示的样子。

In [28]:
annual_2020 = annual[annual["年份"] == 2020]  # 取出 2020 年样本
annual_2020["股票简称"].nunique()  # 计算公司数量

23

In [29]:
annual_2020["营业收入_亿元"].mean()  # 计算营业收入平均值

np.float64(473.61260869565217)

In [30]:
annual_2020["营业收入_亿元"].max()  # 计算营业收入最大值

4191.12

In [31]:
annual_2020[["营业收入_亿元", "净利润_亿元", "资产负债率", "净利率", "人均营业收入_万元"]].describe()  # 查看多列描述统计

,营业收入_亿元,净利润_亿元,资产负债率,净利率,人均营业收入_万元
count,23.00,23.00,23.00,23.00,23.00
mean,473.61,46.05,0.53,0.07,179.67
std,1063.12,133.44,0.20,0.09,157.19
min,1.11,-0.20,0.07,-0.18,16.57
25%,25.22,0.60,0.44,0.02,74.25
50%,47.22,4.66,0.60,0.04,150.24
75%,171.82,8.53,0.68,0.09,214.27
max,4191.12,592.98,0.81,0.31,739.29


In [32]:
rank_display = annual_2020_rank.rename(
    columns={
        "营业收入_亿元": "营业收入",  # 改短列名
        "净利润_亿元": "净利润",  # 改短列名
    }
)

rank_display[["股票简称", "行业名称", "营业收入", "净利润", "净利率"]].head()  # 查看展示表

,股票简称,行业名称,营业收入,净利润,净利率
2,万科A,房地产业,4191.12,592.98,0.14
41,美的集团,电气机械及器材制造业,2842.21,275.07,0.10
44,潍柴动力,汽车制造业,1974.91,112.75,0.06
17,深康佳A,计算机、通信和其他电子设备制造业,503.52,5.40,0.01
56,江铃汽车,汽车制造业,330.96,5.51,0.02


阶段 3 小结：我们得到了 `annual_2020_rank`。这一阶段带出了日期列提取、列运算、新增列、按条件赋值、`np.where()`、`pd.cut()`、排序、排名和简单统计。这张排名表先放在这里；接下来换到另一张指标表，处理它里面更接近真实数据的格式问题，等清洗完成后再和经营表合并。

**练习**：完成最后的阶段 3 练习。

## 阶段 4：常见数据问题处理

真实数据进入统计和合并前，通常要先检查会影响结果的问题。比如证券代码被读成数字会丢掉前导 0，日期列里混入非法日期会影响年份提取，财务指标里出现 `--`、百分号、逗号或“缺失”会影响计算。清洗的目标是让关键列能用于计算、筛选和合并。

范例：读入一张待清洗的公司年度财务指标表，把证券代码整理成 6 位字符串，把统计截止日期整理成日期，把收益率、每股收益和利润指标整理成数值列。处理完成后，这张表要能按“公司 + 日期”和经营表合并。

**读入脏数据，并先看一眼**

1. `converters={"证券代码": str}`：读取 Excel 时把证券代码直接读成字符串。证券代码是编号，不适合作为普通数值处理。

2. `head()` / `tail()`：清洗前先看几行，观察问题大致出现在哪里。

3. `dtypes`：查看每列类型。数字列如果显示为 `object`，常常说明里面混入了文本。看到这种情况，先找出哪些值不像数字，再处理类型。

In [33]:
# 读入待清洗的财务指标表，先保留证券代码的文本形态。
indicators_raw = pd.read_excel(
    "data/financial_indicators_dirty.xlsx",
    converters={"证券代码": str},  # 证券代码按字符串读取
)

indicators_raw.head()  # 查看脏数据前几行

,证券代码,股票简称,统计截止日期,加权平均净资产收益率,扣非净资产收益率,基本每股收益,非经常性损益_亿元,扣非净利润_亿元
0,2,万科A,2018-12-31 00:00:00,23.24,23.05,3.06,2.83,334.90
1,000002,万科A,2019-12-31 00:00:00,22.47,22.14,3.47,5.58,383.14
2,000002,万科A,2020/12/31,20.13,19.51,3.62,12.78,402.38
3,000004,国农科技,2020-13-31,-17,-19.29,-0.24,0.03,-0.23
4,000004,国农科技,2019-12-31 00:00:00,2.97,-24.12,0.04,--,-0.27


In [34]:
indicators_raw.tail()  # 查看脏数据末尾

,证券代码,股票简称,统计截止日期,加权平均净资产收益率,扣非净资产收益率,基本每股收益,非经常性损益_亿元,扣非净利润_亿元
41,000333,美的集团,2020-12-31 00:00:00,24.95,22.56,3.93,26.08,246.15
42,000338,潍柴动力,2018-12-31 00:00:00,22.73,21.02,1.08,6.48,80.09
43,000338,潍柴动力,2019-12-31 00:00:00,21.34,19.58,1.15,7.82,83.23
44,000338,潍柴动力,2020-12-31 00:00:00,19.19,18.02,1.16,4.78,87.29
45,2,万科A,2018-12-31 00:00:00,23.24,23.05,3.06,2.83,334.90


In [35]:
indicators_raw.dtypes  # 查看每列数据类型

证券代码           object
股票简称           object
统计截止日期         object
加权平均净资产收益率     object
扣非净资产收益率      float64
基本每股收益         object
非经常性损益_亿元      object
扣非净利润_亿元      float64
dtype: object

**检查缺失值和重复记录**

1. `isna()`：判断每个位置是否为缺失值。

2. `sum()`：对 `True` / `False` 求和时，`True` 会按 1 计算，因此 `isna().sum()` 可以统计每列缺失值数量。

3. `duplicated(subset=[...])`：按指定列检查重复。公司年度数据通常要检查“公司 + 日期”是否重复。只看重复数量还不够，最好把重复行展示出来，确认重复是否真的应当删除。

In [36]:
indicators_raw.isna().sum()  # 统计每列缺失值数量

证券代码          0
股票简称          0
统计截止日期        0
加权平均净资产收益率    0
扣非净资产收益率      1
基本每股收益        0
非经常性损益_亿元     0
扣非净利润_亿元      0
dtype: int64

In [37]:
indicators_raw.duplicated(subset=["证券代码", "统计截止日期"]).sum()  # 检查公司日期是否重复

np.int64(1)

In [38]:
# 把重复的公司日期记录展示出来，确认问题长什么样。
dup_mask = indicators_raw.duplicated(subset=["证券代码", "统计截止日期"], keep=False)  # 标记所有重复行
indicators_raw.loc[dup_mask, ["证券代码", "股票简称", "统计截止日期"]].sort_values(["证券代码", "统计截止日期"])  # 查看重复记录

,证券代码,股票简称,统计截止日期
0,2,万科A,2018-12-31 00:00:00
45,2,万科A,2018-12-31 00:00:00


**清理编号和日期**

1. `.str.strip()`：去掉文本两端空格。

2. `.str.zfill(6)`：把证券代码补齐到 6 位。

3. `pd.to_datetime(errors="coerce")`：把日期列转换成日期类型；无法识别的日期会变成缺失值。转换后如果缺失值增加，通常说明原始日期里有无法识别的写法。

4. `.dt.year`：从日期列中提取年份。

In [39]:
# 先整理公司编号和公司简称。
indicators = indicators_raw.copy()  # 复制工作副本
indicators["证券代码"] = indicators["证券代码"].str.strip().str.zfill(6)  # 去空格并补齐 6 位
indicators["股票简称"] = indicators["股票简称"].str.strip()  # 去掉简称两端空格

indicators[["证券代码", "股票简称", "统计截止日期"]].head(8)  # 查看编号和简称

,证券代码,股票简称,统计截止日期
0,000002,万科A,2018-12-31 00:00:00
1,000002,万科A,2019-12-31 00:00:00
2,000002,万科A,2020/12/31
3,000004,国农科技,2020-13-31
4,000004,国农科技,2019-12-31 00:00:00
5,000004,国华网安,2020-12-31 00:00:00
6,000006,深振业A,2018-12-31 00:00:00
7,000006,深振业A,2019-12-31 00:00:00


In [40]:
# 再把统计截止日期转换成日期，并提取年份。
indicators["统计截止日期"] = pd.to_datetime(indicators["统计截止日期"], errors="coerce")  # 转换日期
indicators["年份"] = indicators["统计截止日期"].dt.year  # 提取年份

indicators[["证券代码", "股票简称", "统计截止日期", "年份"]].head(8)  # 查看日期处理结果

,证券代码,股票简称,统计截止日期,年份
0,000002,万科A,2018-12-31,2018.00
1,000002,万科A,2019-12-31,2019.00
2,000002,万科A,2020-12-31,2020.00
3,000004,国农科技,NaT,NaN
4,000004,国农科技,2019-12-31,2019.00
5,000004,国华网安,2020-12-31,2020.00
6,000006,深振业A,2018-12-31,2018.00
7,000006,深振业A,2019-12-31,2019.00


In [41]:
indicators["统计截止日期"].isna().sum()  # 检查无法识别的日期数量

np.int64(1)

In [42]:
indicators.loc[indicators["统计截止日期"].isna(), ["证券代码", "股票简称", "年份"]]  # 查看日期转换失败的记录

,证券代码,股票简称,年份
3,000004,国农科技,NaN


**清理数值列中的特殊文本**

1. `.astype("string")`：先把混杂列统一看作文本，便于使用字符串方法清理。

2. `.str.replace()`：去掉逗号、百分号等字符。

3. `.replace({...})`：把 `--`、`缺失`、空字符串等特殊编码替换成缺失值。

4. `pd.to_numeric(errors="coerce")`：把清理后的文本转成数值；仍无法转换的内容会变成缺失值。这个参数很适合排查脏值：先转换，再回头看哪些原始值变成了缺失。

In [43]:
# 先示范清理一列：加权平均净资产收益率。
roe_before = indicators["加权平均净资产收益率"].copy()  # 保存清理前结果
indicators["加权平均净资产收益率"] = pd.to_numeric(
    indicators["加权平均净资产收益率"]
    .astype("string")  # 统一转成字符串
    .str.strip()  # 去掉两端空格
    .str.replace("%", "", regex=False),  # 去掉百分号
    errors="coerce",  # 无法转换的内容变成缺失值
)

pd.DataFrame({
    "证券代码": indicators["证券代码"],
    "股票简称": indicators["股票简称"],
    "清理前": roe_before,
    "清理后": indicators["加权平均净资产收益率"],
}).head(8)  # 对比清理前后

,证券代码,股票简称,清理前,清理后
0,000002,万科A,23.24,23.24
1,000002,万科A,22.47,22.47
2,000002,万科A,20.13,20.13
3,000004,国农科技,-17,-17.00
4,000004,国农科技,2.97,2.97
5,000004,国华网安,4.54,4.54
6,000006,深振业A,14.86%,14.86
7,000006,深振业A,12.39,12.39


In [44]:
# 找出原来不是缺失、转换后变成缺失的记录。
roe_failed = roe_before.notna() & indicators["加权平均净资产收益率"].isna()  # 标记转换失败
pd.DataFrame({
    "证券代码": indicators.loc[roe_failed, "证券代码"],
    "股票简称": indicators.loc[roe_failed, "股票简称"],
    "原始取值": roe_before.loc[roe_failed],
})  # 查看无法转换的原始值

,证券代码,股票简称,原始取值


In [45]:
# 再用同样规则批量清理其他数值列。
def clean_number(s):
    return pd.to_numeric(
        s.astype("string")  # 统一转成字符串
        .str.strip()  # 去掉两端空格
        .str.replace(",", "", regex=False)  # 去掉千分位逗号
        .str.replace("%", "", regex=False)  # 去掉百分号
        .replace({"--": pd.NA, "缺失": pd.NA, "": pd.NA}),  # 特殊文本改成缺失
        errors="coerce",  # 无法转换的内容变成缺失值
    )

number_cols = [
    "加权平均净资产收益率",
    "扣非净资产收益率",
    "基本每股收益",
    "非经常性损益_亿元",
    "扣非净利润_亿元",
]

for col in number_cols[1:]:
    indicators[col] = clean_number(indicators[col])  # 批量清理其余数值列

indicators[["证券代码", "股票简称", "统计截止日期"] + number_cols].head(8)  # 查看数值转换结果

,证券代码,股票简称,统计截止日期,加权平均净资产收益率,扣非净资产收益率,基本每股收益,非经常性损益_亿元,扣非净利润_亿元
0,000002,万科A,2018-12-31,23.24,23.05,3.06,2.83,334.90
1,000002,万科A,2019-12-31,22.47,22.14,3.47,5.58,383.14
2,000002,万科A,2020-12-31,20.13,19.51,3.62,12.78,402.38
3,000004,国农科技,NaT,-17.00,-19.29,-0.24,0.03,-0.23
4,000004,国农科技,2019-12-31,2.97,-24.12,0.04,<NA>,-0.27
5,000004,国华网安,2020-12-31,4.54,4.42,0.39,0.02,0.63
6,000006,深振业A,2018-12-31,14.86,14.80,0.65,0.04,8.71
7,000006,深振业A,2019-12-31,12.39,<NA>,0.59,0.22,7.79


In [46]:
indicators[number_cols].dtypes  # 检查清理后的数值类型

加权平均净资产收益率    Float64
扣非净资产收益率      Float64
基本每股收益        Float64
非经常性损益_亿元     Float64
扣非净利润_亿元      Float64
dtype: object

**删除重复记录和关键缺失**

1. `drop_duplicates(subset=[...])`：按指定列删除重复记录。

2. `dropna(subset=[...])`：只检查关键列，删除这些列中有缺失值的行。

3. `fillna()`：用指定值填补缺失值。是否填补要看变量含义；这里用每股收益的中位数做演示。连接键缺失通常不适合填补，指标缺失才可能根据业务含义选择填补。

In [47]:
# 删除重复记录和关键连接键缺失的记录。
indicators_clean = (
    indicators
    .drop_duplicates(subset=["证券代码", "统计截止日期"])  # 删除重复的公司日期记录
    .dropna(subset=["证券代码", "统计截止日期"])  # 删除连接键缺失的记录
    .copy()
)

len(indicators_clean)  # 查看清理后的行数

44

In [48]:
# 对少量指标缺失做填补，并生成一个便于观察的标签。
indicators_clean["基本每股收益"] = indicators_clean["基本每股收益"].fillna(
    indicators_clean["基本每股收益"].median()  # 用中位数填补
)
indicators_clean["扣非盈利"] = indicators_clean["扣非净利润_亿元"] > 0  # 生成布尔标签

indicators_clean[["证券代码", "股票简称", "统计截止日期", "基本每股收益", "扣非净利润_亿元", "扣非盈利"]].head()  # 查看清理结果

,证券代码,股票简称,统计截止日期,基本每股收益,扣非净利润_亿元,扣非盈利
0,000002,万科A,2018-12-31,3.06,334.90,True
1,000002,万科A,2019-12-31,3.47,383.14,True
2,000002,万科A,2020-12-31,3.62,402.38,True
4,000004,国农科技,2019-12-31,0.04,-0.27,False
5,000004,国华网安,2020-12-31,0.39,0.63,True


In [49]:
indicators_clean.isna().sum()  # 检查清理后的缺失值

证券代码          0
股票简称          0
统计截止日期        0
加权平均净资产收益率    0
扣非净资产收益率      1
基本每股收益        0
非经常性损益_亿元     1
扣非净利润_亿元      0
年份            0
扣非盈利          0
dtype: int64

### 副本、视图和链式赋值

直接修改数据时，建议优先使用：

```python
df.loc[条件, 列名] = 新值
```

筛选出一部分数据再修改时，显式使用 `.copy()` 生成工作副本。这样更容易判断后续操作影响的是哪一张表。连续使用 `df[条件]["列"] = 新值` 这类链式赋值，pandas 往往会给出 warning，也可能没有改到原表。

阶段 4 小结：我们得到了 `indicators_clean`。这一阶段带出了 `converters`、`isna()`、`duplicated()`、`drop_duplicates()`、`str.strip()`、`zfill()`、`pd.to_datetime()`、`pd.to_numeric()`、`replace()`、`dropna()`、`fillna()`、工作副本和链式赋值 warning。清洗后要看三件事：连接键是否可用，数值列是否已经是数值类型，行数变化是否能解释。

**练习**：完成最后的阶段 4 练习。

## 阶段 5：表的追加与合并

很多分析表不是一次拿到的，而是由多张表拼出来的。同一结构的数据可以上下追加；不同信息的数据需要按连接键合并。阶段 4 已经把财务指标表整理成可合并的形态，这里让它和经营表碰面，说明 `concat()` 和 `merge()` 的基本用法。

范例：先把 2018-2019 年和 2020 年的经营记录纵向追加，再把清洗后的财务指标合并进去，得到后续分析使用的 `analysis_df`。

**纵向追加同结构表**

1. `pd.concat([表1, 表2, ...], ignore_index=True)`：把列结构相同或相近的表纵向接在一起。

2. `ignore_index=True`：拼接后重新生成连续索引。追加多张同结构表时，通常使用这个参数。

3. `concat()` 和 `merge()` 解决的问题不同：`concat()` 是把记录追加到一起，`merge()` 是按连接键把不同信息匹配到同一行。

In [50]:
# 重新读入带证券代码的经营表，读完先看面板核心列和关键经营列。
def read_code(x):
    return str(x).strip().zfill(6)  # 转字符串、去空格、补齐 6 位

annual_for_merge = pd.read_excel(
    "data/company_annual_operations_clean.xlsx",
    converters={"证券代码": read_code},  # 证券代码按字符串读取
)
annual_for_merge = annual_for_merge.rename(columns={"股票简称": "证券简称"})  # 后续阶段沿用证券简称

annual_for_merge[["证券代码", "证券简称", "统计截止日期", "营业收入_亿元", "净利润_亿元", "资产负债率"]].head()  # 查看经营表

,证券代码,证券简称,统计截止日期,营业收入_亿元,净利润_亿元,资产负债率
0,000002,万科A,2018-12-31,2976.79,492.72,0.85
1,000002,万科A,2019-12-31,3678.94,551.32,0.84
2,000002,万科A,2020-12-31,4191.12,592.98,0.81
3,000004,国农科技,2018-12-31,3.67,-0.22,0.48
4,000004,国农科技,2019-12-31,1.08,-0.04,0.06


In [51]:
# 补充后面分组和合并需要的变量。
annual_for_merge["年份"] = annual_for_merge["统计截止日期"].dt.year  # 提取年份
annual_for_merge["上市年份"] = annual_for_merge["首次上市日期"].dt.year  # 提取上市年份
annual_for_merge["上市年限"] = annual_for_merge["年份"] - annual_for_merge["上市年份"]  # 计算上市年限
annual_for_merge["净利率"] = annual_for_merge["净利润_亿元"] / annual_for_merge["营业收入_亿元"]  # 计算净利率
annual_for_merge["负债水平"] = np.where(annual_for_merge["资产负债率"] >= 0.7, "较高", "正常")  # 生成负债水平

annual_for_merge[["证券代码", "证券简称", "年份", "上市年限", "净利率", "负债水平"]].head()  # 查看新增变量

,证券代码,证券简称,年份,上市年限,净利率,负债水平
0,000002,万科A,2018,27,0.17,较高
1,000002,万科A,2019,28,0.15,较高
2,000002,万科A,2020,29,0.14,较高
3,000004,国农科技,2018,27,-0.06,正常
4,000004,国农科技,2019,28,-0.04,正常


In [52]:
# 把两个时间段的经营记录纵向追加到一起。
annual_2018_2019 = annual_for_merge.loc[annual_for_merge["年份"].isin([2018, 2019])].copy()  # 2018-2019 年
annual_2020_part = annual_for_merge.loc[annual_for_merge["年份"] == 2020].copy()  # 2020 年

annual_panel = pd.concat(
    [annual_2018_2019, annual_2020_part],  # 要追加的表
    ignore_index=True,  # 重新生成连续索引
)

annual_panel[["证券代码", "证券简称", "年份", "营业收入_亿元"]].head()  # 查看追加结果

,证券代码,证券简称,年份,营业收入_亿元
0,000002,万科A,2018,2976.79
1,000002,万科A,2019,3678.94
2,000004,国农科技,2018,3.67
3,000004,国农科技,2019,1.08
4,000006,深振业A,2018,25.12


In [53]:
annual_panel["年份"].value_counts().sort_index()  # 检查追加后的年度记录数

年份
2018    23
2019    23
2020    23
Name: count, dtype: int64

**按连接键合并表**

1. `merge()`：按共同字段合并两张表。

2. `on=[...]`：指定一列或多列作为连接键。公司年度数据通常需要同时使用公司代码和日期。

3. `how="left"`：保留左表所有行，把右表能匹配的信息接上来。

4. 合并后检查：通常检查合并前后行数、关键列缺失数量，并看几行结果。`left` 合并后行数变多，常见原因是右表连接键重复；`inner` 合并后行数变少，说明有些左表记录在右表中匹配不上。

In [54]:
# 把清洗后的财务指标合并到经营表。
indicator_cols = [
    "证券代码",
    "统计截止日期",
    "加权平均净资产收益率",
    "扣非净资产收益率",
    "基本每股收益",
    "非经常性损益_亿元",
    "扣非净利润_亿元",
    "扣非盈利",
]

analysis_df = annual_panel.merge(
    indicators_clean[indicator_cols],
    on=["证券代码", "统计截止日期"],  # 按公司和日期匹配
    how="left",  # 保留经营表所有行
)

analysis_df[["证券代码", "证券简称", "年份", "营业收入_亿元", "基本每股收益", "扣非净利润_亿元"]].head(10)  # 查看合并结果

,证券代码,证券简称,年份,营业收入_亿元,基本每股收益,扣非净利润_亿元
0,000002,万科A,2018,2976.79,3.06,334.90
1,000002,万科A,2019,3678.94,3.47,383.14
2,000004,国农科技,2018,3.67,<NA>,<NA>
3,000004,国农科技,2019,1.08,0.04,-0.27
4,000006,深振业A,2018,25.12,0.65,8.71
5,000006,深振业A,2019,37.31,0.59,7.79
6,000011,深物业A,2018,27.87,0.99,5.91
7,000011,深物业A,2019,39.62,1.37,6.96
8,000014,沙河股份,2018,3.58,0.72,0.05
9,000014,沙河股份,2019,4.38,0.10,0.12


In [55]:
indicators_clean.duplicated(subset=["证券代码", "统计截止日期"]).sum()  # 合并前检查右表连接键是否重复

np.int64(0)

In [56]:
len(annual_panel)  # 合并前行数

69

In [57]:
len(analysis_df)  # 合并后行数

69

In [58]:
analysis_df["基本每股收益"].isna().sum()  # 检查右表指标缺失数量

np.int64(25)

**不同合并方式的行数**

1. `how="left"`：保留左表中的行。

2. `how="inner"`：只保留两表都能匹配的行。

3. `how="outer"`：保留两表行的并集。

In [59]:
# 比较不同 how 参数得到的行数。
merge_how_demo = pd.DataFrame({
    "how": ["left", "inner", "outer"],
    "行数": [
        len(annual_panel.merge(indicators_clean[indicator_cols], on=["证券代码", "统计截止日期"], how="left")),  # left 合并行数
        len(annual_panel.merge(indicators_clean[indicator_cols], on=["证券代码", "统计截止日期"], how="inner")),  # inner 合并行数
        len(annual_panel.merge(indicators_clean[indicator_cols], on=["证券代码", "统计截止日期"], how="outer")),  # outer 合并行数
    ],
})

merge_how_demo  # 查看不同合并方式的行数

,how,行数
0,left,69
1,inner,44
2,outer,69


阶段 5 小结：我们得到了 `analysis_df`。这一阶段带出了 `concat()`、`ignore_index`、`merge()`、多列连接键、`how` 和合并检查。合并后的表是后面分组、重塑和时间顺序计算的共同起点；以后遇到合并结果异常，先查连接键和行数。

**练习**：完成最后的阶段 5 练习。

## 阶段 6：分组汇总与组内比较

公司年度明细表是一家公司一年一行。很多经管问题关心的不是单家公司，而是行业、年份或组内位置。比如，2020 年哪些行业的收入规模更大，不同行业每年的平均净利率如何，某家公司是否高于本行业平均水平。`analysis_df` 已经把经营数据和指标数据接到一起，可以开始从明细表走向分组结论。

范例：从 `analysis_df` 出发，先看 2020 年各行业收入规模和盈利情况，再把行业平均收入带回公司明细表，判断每家公司是否高于本行业平均水平。

In [60]:
analysis_df[[
    "证券代码", "证券简称", "年份", "行业名称", "营业收入_亿元", "净利率", "资产负债率", "基本每股收益"
]].head()  # 查看合并后的明细表

,证券代码,证券简称,年份,行业名称,营业收入_亿元,净利率,资产负债率,基本每股收益
0,000002,万科A,2018,房地产业,2976.79,0.17,0.85,3.06
1,000002,万科A,2019,房地产业,3678.94,0.15,0.84,3.47
2,000004,国农科技,2018,医药制造业,3.67,-0.06,0.48,<NA>
3,000004,国农科技,2019,软件和信息技术服务业,1.08,-0.04,0.06,0.04
4,000006,深振业A,2018,房地产业,25.12,0.37,0.53,0.65


**按一个分组列汇总**

1. `groupby("分组列")`：按照某一列把数据分组。

2. `.agg(新列名=("被计算列", "统计方法"))`：对每组计算统计量，并给结果列命名。

3. `nunique()`：计算不重复数量。这里用来统计每个行业有多少家公司。

4. `sort_values()`：按汇总结果排序，便于先看最重要的组。

分组汇总后要看分组数量和样本量。只有一两家公司的行业，均值和最大值可能更像个案，不一定代表行业整体。

In [61]:
# 先取出 2020 年公司明细，后面围绕这一年做行业比较。
analysis_2020 = analysis_df.loc[analysis_df["年份"] == 2020].copy()  # 筛选 2020 年

analysis_2020[["证券代码", "证券简称", "行业名称", "营业收入_亿元", "净利率", "资产负债率"]].head()  # 查看 2020 年明细

,证券代码,证券简称,行业名称,营业收入_亿元,净利率,资产负债率
46,000002,万科A,房地产业,4191.12,0.14,0.81
47,000004,国华网安,软件和信息技术服务业,2.81,0.22,0.07
48,000006,深振业A,房地产业,29.35,0.31,0.49
49,000011,深物业A,房地产业,41.04,0.18,0.69
50,000014,沙河股份,房地产业,3.47,0.03,0.61


In [62]:
# 按行业汇总 2020 年经营表现。
industry_2020 = (
    analysis_2020
    .groupby("行业名称")  # 按行业分组
    .agg(
        公司数=("证券简称", "nunique"),  # 每个行业公司数量
        平均营业收入_亿元=("营业收入_亿元", "mean"),  # 行业平均收入
        营业收入合计_亿元=("营业收入_亿元", "sum"),  # 行业收入合计
        平均净利率=("净利率", "mean"),  # 行业平均净利率
        平均资产负债率=("资产负债率", "mean"),  # 行业平均资产负债率
    )
    .sort_values("营业收入合计_亿元", ascending=False)  # 按收入合计降序
)

industry_2020  # 查看行业汇总结果

,公司数,平均营业收入_亿元,营业收入合计_亿元,平均净利率,平均资产负债率
行业名称,,,,,
房地产业,4,1066.24,4264.98,0.16,0.65
电气机械及器材制造业,4,798.83,3195.32,0.05,0.63
汽车制造业,4,631.46,2525.82,0.05,0.54
计算机、通信和其他电子设备制造业,4,170.30,681.20,0.03,0.52
软件和信息技术服务业,4,36.99,147.97,0.04,0.43
医药制造业,3,25.93,77.80,0.07,0.37


**按多个分组列汇总**

1. `groupby(["列1", "列2"])`：同时按多个列分组。这里用行业和年份，得到行业-年份层面的结果。

2. `reset_index()`：把分组结果中的索引还原成普通列，便于继续筛选、排序或保存。

In [63]:
# 同时按行业和年份汇总，观察行业指标随年份变化。
industry_year = (
    analysis_df
    .groupby(["行业名称", "年份"])  # 按行业和年份分组
    .agg(
        公司数=("证券简称", "nunique"),
        平均营业收入_亿元=("营业收入_亿元", "mean"),
        平均净利率=("净利率", "mean"),
    )
    .reset_index()  # 把行业和年份恢复成普通列
)

industry_year.head(12)  # 查看行业-年份汇总

,行业名称,年份,公司数,平均营业收入_亿元,平均净利率
0,医药制造业,2018,4,28.95,0.08
1,医药制造业,2019,3,23.71,0.02
2,医药制造业,2020,3,25.93,0.07
3,房地产业,2018,4,758.34,0.29
4,房地产业,2019,4,940.06,0.16
5,房地产业,2020,4,1066.24,0.16
6,汽车制造业,2018,4,516.80,0.06
7,汽车制造业,2019,4,560.45,0.05
8,汽车制造业,2020,4,631.46,0.05
9,电气机械及器材制造业,2018,5,587.76,0.05


**自定义聚合函数**

1. 自定义函数可以放进 `agg()`。函数接收一列数据，返回一个统计结果。

2. 对分组后的结果继续 `sort_values()`，可以直接找到差异较大的组。

In [64]:
def value_range(x):
    return x.max() - x.min()  # 最大值减最小值

value_range(analysis_2020["营业收入_亿元"])  # 先用一列数据试一下函数

4190.01

In [65]:
# 计算 2020 年各行业收入均值和行业内收入差距。
industry_gap_2020 = (
    analysis_2020
    .groupby("行业名称")
    .agg(
        收入均值=("营业收入_亿元", "mean"),
        收入差距=("营业收入_亿元", value_range),
    )
    .sort_values("收入差距", ascending=False)
)

industry_gap_2020.head()  # 查看收入差距较大的行业

,收入均值,收入差距
行业名称,,
房地产业,1066.24,4187.65
电气机械及器材制造业,798.83,2794.99
汽车制造业,631.46,1866.09
计算机、通信和其他电子设备制造业,170.30,496.60
软件和信息技术服务业,36.99,105.75


**把分组结果带回明细表**

1. `transform()`：把每组的计算结果放回每一行。`agg()` 通常得到每组一行，`transform()` 得到和原表等长的一列。

2. 组内比较常用 `transform()`。例如先计算行业平均收入，再判断每家公司是否高于本行业平均水平。

一个简单判断标准：如果结果只需要每组一行，用 `agg()`；如果结果还要和原表每一行放在一起继续比较，用 `transform()`。

In [66]:
# 计算每家公司所在行业的平均收入，并判断是否高于行业平均。
industry_compare_2020 = analysis_2020.copy()  # 复制 2020 年明细表
industry_compare_2020["行业平均收入_亿元"] = (
    industry_compare_2020
    .groupby("行业名称")["营业收入_亿元"]
    .transform("mean")  # 把行业平均值带回每一行
)
industry_compare_2020["高于行业平均"] = (
    industry_compare_2020["营业收入_亿元"] > industry_compare_2020["行业平均收入_亿元"]
)

industry_compare_2020[[
    "行业名称", "证券简称", "营业收入_亿元", "行业平均收入_亿元", "高于行业平均"
]].head(10)  # 查看组内比较结果

,行业名称,证券简称,营业收入_亿元,行业平均收入_亿元,高于行业平均
46,房地产业,万科A,4191.12,1066.24,True
47,软件和信息技术服务业,国华网安,2.81,36.99,False
48,房地产业,深振业A,29.35,1066.24,False
49,房地产业,深物业A,41.04,1066.24,False
50,房地产业,沙河股份,3.47,1066.24,False
51,计算机、通信和其他电子设备制造业,深康佳A,503.52,170.30,True
52,计算机、通信和其他电子设备制造业,深华发A,6.92,170.30,False
53,计算机、通信和其他电子设备制造业,深科技,149.67,170.30,False
54,汽车制造业,富奥股份,111.13,631.46,False
55,计算机、通信和其他电子设备制造业,深纺织A,21.09,170.30,False


**每组取前几名**

1. 常用步骤是先排序，再 `groupby()`，最后 `head(n)`。

2. 排序时先按分组列排，再按组内比较指标排，可以让结果更容易检查。

In [67]:
# 每个行业取营业收入最高的 2 家公司。
top2_by_industry = (
    analysis_2020
    .sort_values(["行业名称", "营业收入_亿元"], ascending=[True, False])  # 先行业，再收入降序
    .groupby("行业名称")
    .head(2)  # 每组取前 2 行
)

top2_by_industry[["行业名称", "证券代码", "证券简称", "营业收入_亿元", "净利率"]].head(14)  # 查看每组前 2 名

,行业名称,证券代码,证券简称,营业收入_亿元,净利率
63,医药制造业,000423,东阿阿胶,34.09,0.01
58,医药制造业,000153,丰原药业,33.21,0.03
46,房地产业,000002,万科A,4191.12,0.14
49,房地产业,000011,深物业A,41.04,0.18
60,汽车制造业,000338,潍柴动力,1974.91,0.06
64,汽车制造业,000550,江铃汽车,330.96,0.02
59,电气机械及器材制造业,000333,美的集团,2842.21,0.10
56,电气机械及器材制造业,000049,德赛电池,193.98,0.04
51,计算机、通信和其他电子设备制造业,000016,深康佳A,503.52,0.01
53,计算机、通信和其他电子设备制造业,000021,深科技,149.67,0.06


**分组循环**

1. `for 组名, 小表 in df.groupby("分组列")`：逐组取出数据。组内逻辑比较复杂时，循环比一行链式写法更容易读。

2. 循环得到多个小表后，可以用前面学过的 `pd.concat()` 追加回来。

In [68]:
# 用分组循环找出每个行业收入最高的 1 家公司。
top_list = []

for industry, group in analysis_2020.groupby("行业名称"):
    top_company = group.sort_values("营业收入_亿元", ascending=False).head(1)  # 本行业收入最高公司
    top_list.append(top_company)

industry_top_company = pd.concat(top_list, ignore_index=True)[
    ["行业名称", "证券代码", "证券简称", "营业收入_亿元", "净利率"]
].sort_values("营业收入_亿元", ascending=False)

industry_top_company  # 查看每个行业收入最高的公司

,行业名称,证券代码,证券简称,营业收入_亿元,净利率
1,房地产业,000002,万科A,4191.12,0.14
3,电气机械及器材制造业,000333,美的集团,2842.21,0.10
2,汽车制造业,000338,潍柴动力,1974.91,0.06
4,计算机、通信和其他电子设备制造业,000016,深康佳A,503.52,0.01
5,软件和信息技术服务业,000555,神州信息,106.86,0.04
0,医药制造业,000423,东阿阿胶,34.09,0.01


阶段 6 小结：这一阶段带出了 `groupby()`、多指标 `agg()`、自定义聚合、`transform()`、每组取前几名、分组循环和在循环结果中复用 `concat()`。分组结果出来后，既要看排序靠前的组，也要看每组公司数，避免把样本很少的组解读得太重。

**练习**：完成最后的阶段 6 练习。

## 阶段 7：数据重塑与结果表构造

明细表适合记录原始观测，但不一定适合直接回答问题。比如，如果想比较每家公司 2018 年到 2020 年的营业收入变化，把年份展开成列会更直观；如果要把结果发给别人，也需要整理成一张结构清楚的结果表。阶段 6 得到的是行业层面的汇总，阶段 7 回到公司层面，构造一张便于展示和保存的结果表。

范例：把公司年度营业收入从长表整理成宽表，每家公司一行，2018、2019、2020 年收入分别成为列；再计算 2018-2020 年收入增长率，并整理成可以保存的公司结果表。

**长表和宽表**

1. 长表：同一个变量的不同年份放在多行里。当前 `analysis_df` 就是长表，一家公司一年一行。

2. 宽表：同一个变量的不同年份展开成多列。比较跨年变化时，宽表更方便。

3. `pivot_table()`：把长表整理成宽表。需要指定行索引、展开成列的变量，以及单元格里的数值。重塑后要检查行数和年份列，确认没有把公司合并错，也没有漏掉年份。

In [69]:
# 先看用于重塑的明细列。
analysis_df[["证券代码", "证券简称", "行业名称", "年份", "营业收入_亿元", "资产负债率"]].head(9)  # 查看长表

,证券代码,证券简称,行业名称,年份,营业收入_亿元,资产负债率
0,000002,万科A,房地产业,2018,2976.79,0.85
1,000002,万科A,房地产业,2019,3678.94,0.84
2,000004,国农科技,医药制造业,2018,3.67,0.48
3,000004,国农科技,软件和信息技术服务业,2019,1.08,0.06
4,000006,深振业A,房地产业,2018,25.12,0.53
5,000006,深振业A,房地产业,2019,37.31,0.54
6,000011,深物业A,房地产业,2018,27.87,0.43
7,000011,深物业A,房地产业,2019,39.62,0.70
8,000014,沙河股份,房地产业,2018,3.58,0.50


In [70]:
# 把公司年度营业收入整理成宽表。
revenue_wide = analysis_df.pivot_table(
    index=["证券代码", "证券简称", "行业名称"],  # 每家公司一行
    columns="年份",  # 年份展开成列
    values="营业收入_亿元",  # 单元格里放营业收入
)

revenue_wide.head()  # 查看宽表

年份                        2018    2019    2020
证券代码   证券简称 行业名称                              
000002 万科A  房地产业       2976.79 3678.94 4191.12
000004 国农科技 医药制造业         3.67     NaN     NaN
            软件和信息技术服务业     NaN    1.08     NaN
       国华网安 软件和信息技术服务业     NaN     NaN    2.81
000006 深振业A 房地产业         25.12   37.31   29.35

**在宽表上继续计算**

1. 宽表生成后，不同年份已经是不同列，可以直接做列运算。

2. `reset_index()`：把索引还原成普通列。结果表通常更适合把公司代码、简称、行业作为普通列展示。

宽表上的列运算很直观，但也要看分母是否缺失或为 0。增长率异常大时，先回到原始收入列看原因。

In [71]:
# 在宽表上计算 2018 到 2020 年营业收入增长率。
revenue_wide["收入增长率_2018_2020"] = revenue_wide[2020] / revenue_wide[2018] - 1  # 跨年增长率

revenue_wide[[2018, 2019, 2020, "收入增长率_2018_2020"]].head()  # 查看新增列

年份                        2018    2019    2020  收入增长率_2018_2020
证券代码   证券简称 行业名称                                               
000002 万科A  房地产业       2976.79 3678.94 4191.12             0.41
000004 国农科技 医药制造业         3.67     NaN     NaN              NaN
            软件和信息技术服务业     NaN    1.08     NaN              NaN
       国华网安 软件和信息技术服务业     NaN     NaN    2.81              NaN
000006 深振业A 房地产业         25.12   37.31   29.35             0.17

In [72]:
# 把索引还原成普通列，并按增长率排序。
company_summary = (
    revenue_wide
    .reset_index()  # 公司代码、简称、行业回到普通列
    .sort_values("收入增长率_2018_2020", ascending=False)  # 按增长率降序
)

company_summary[["证券代码", "证券简称", "行业名称", 2018, 2020, "收入增长率_2018_2020"]].head(10)  # 查看结果表

年份,证券代码,证券简称,行业名称,2018,2020,收入增长率_2018_2020
11,000045,深纺织A,计算机、通信和其他电子设备制造业,12.72,21.09,0.66
5,000011,深物业A,房地产业,27.87,41.04,0.47
10,000030,富奥股份,汽车制造业,78.53,111.13,0.42
0,000002,万科A,房地产业,2976.79,4191.12,0.41
17,000400,许继电气,电气机械及器材制造业,82.17,111.91,0.36
16,000338,潍柴动力,汽车制造业,1592.56,1974.91,0.24
23,000555,神州信息,软件和信息技术服务业,90.77,106.86,0.18
22,000550,江铃汽车,汽车制造业,282.49,330.96,0.17
4,000006,深振业A,房地产业,25.12,29.35,0.17
12,000049,德赛电池,电气机械及器材制造业,172.49,193.98,0.12


**索引和普通列之间切换**

1. `set_index("列名")`：把普通列设为索引。需要按某个标识查找行时很方便。

2. `reset_index()`：把索引还原成普通列。保存或继续合并前经常会这样做。

In [73]:
indexed = company_summary.set_index("证券代码")  # 把证券代码设为索引
indexed[["证券简称", "行业名称", 2018, 2020, "收入增长率_2018_2020"]].head()  # 查看索引表

年份,证券简称,行业名称,2018,2020,收入增长率_2018_2020
证券代码,,,,,
000045,深纺织A,计算机、通信和其他电子设备制造业,12.72,21.09,0.66
000011,深物业A,房地产业,27.87,41.04,0.47
000030,富奥股份,汽车制造业,78.53,111.13,0.42
000002,万科A,房地产业,2976.79,4191.12,0.41
000400,许继电气,电气机械及器材制造业,82.17,111.91,0.36


In [74]:
indexed.reset_index()[["证券代码", "证券简称", "行业名称", 2018, 2020, "收入增长率_2018_2020"]].head()  # 索引还原成普通列

年份,证券代码,证券简称,行业名称,2018,2020,收入增长率_2018_2020
0,000045,深纺织A,计算机、通信和其他电子设备制造业,12.72,21.09,0.66
1,000011,深物业A,房地产业,27.87,41.04,0.47
2,000030,富奥股份,汽车制造业,78.53,111.13,0.42
3,000002,万科A,房地产业,2976.79,4191.12,0.41
4,000400,许继电气,电气机械及器材制造业,82.17,111.91,0.36


**构造小表**

1. `pd.Series()`：构造一列带索引的数据。

2. `pd.DataFrame({...})`：用字典构造表。字典的键会成为列名，值会成为列数据。

构造小表常用于临时整理规则、演示结果或手工输入少量信息。

In [75]:
s = pd.Series([10, 20, 30], index=["A", "B", "C"], name="得分")  # 构造 Series
s  # 查看 Series

A    10
B    20
C    30
Name: 得分, dtype: int64

In [76]:
demo_df = pd.DataFrame({
    "公司": ["甲", "乙", "丙"],
    "收入": [100, 120, 80],
})  # 构造 DataFrame

demo_df  # 查看小表

,公司,收入
0,甲,100
1,乙,120
2,丙,80


**保存结果**

1. `to_excel("路径", index=False)`：保存为 Excel。适合给人直接打开查看。

2. `to_csv("路径", index=False)`：保存为 CSV。格式简单，通用性强。

3. `index=False`：不把索引写入文件。普通结果表通常使用这个参数。

In [77]:
# 保存前先看要输出的公司结果表。
company_summary[["证券代码", "证券简称", "行业名称", 2018, 2020, "收入增长率_2018_2020"]].head()  # 查看保存对象

年份,证券代码,证券简称,行业名称,2018,2020,收入增长率_2018_2020
11,000045,深纺织A,计算机、通信和其他电子设备制造业,12.72,21.09,0.66
5,000011,深物业A,房地产业,27.87,41.04,0.47
10,000030,富奥股份,汽车制造业,78.53,111.13,0.42
0,000002,万科A,房地产业,2976.79,4191.12,0.41
17,000400,许继电气,电气机械及器材制造业,82.17,111.91,0.36


In [78]:
# 保存几个阶段成果。
analysis_df[analysis_df["年份"] == 2020].to_excel("data/finance_2020_rank_full.xlsx", index=False)  # 保存 2020 年明细
industry_2020.to_excel("data/industry_2020_full_summary.xlsx")  # 保存行业汇总
company_summary.to_excel("data/company_full_summary.xlsx", index=False)  # 保存公司结果表
company_summary.to_csv("data/company_full_summary.csv", index=False)  # 保存 CSV 版本

print("已保存阶段成果")

已保存阶段成果


In [79]:
pd.read_csv("data/company_full_summary.csv").head()  # 读回 CSV 检查

,证券代码,证券简称,行业名称,2018,2019,2020,收入增长率_2018_2020
0,45,深纺织A,计算机、通信和其他电子设备制造业,12.72,21.58,21.09,0.66
1,11,深物业A,房地产业,27.87,39.62,41.04,0.47
2,30,富奥股份,汽车制造业,78.53,100.64,111.13,0.42
3,2,万科A,房地产业,2976.79,3678.94,4191.12,0.41
4,400,许继电气,电气机械及器材制造业,82.17,101.56,111.91,0.36


阶段 7 小结：这一阶段带出了长表和宽表、`pivot_table()`、列运算、`set_index()`、`reset_index()`、构造 `Series` / `DataFrame`，以及保存 Excel 和 CSV。结果表保存前先看几行，保存后再读回来看一眼，这能发现路径、索引和列名问题。

**练习**：完成最后的阶段 7 练习。

## 阶段 8：公司年度面板中的时间顺序

本章的公司经营表已经是一张年度面板：同一家公司在不同年份有多行记录。处理这类数据时，关键是先确认时间顺序，再在公司内部计算上一年、本年变化和增长率。阶段 7 把年份展开成了列；阶段 8 保持长表形态，用时间顺序直接计算变化。

范例：先选出万科 A 的年度经营记录，确认上一年收入、收入增加额和收入增长率怎么算；再把同样的计算扩展到所有公司，找出 2020 年收入增长率较高的公司。

**先确认时间顺序**

1. `sort_values()`：按公司和日期排序。时间序列计算通常要先排序。

2. 同一家公司跨年比较时，先单独取出一家公司看清楚，确认结果符合直觉，再推广到全部公司。

时间计算最怕顺序错。`shift()` 不会自动理解“上一年”，它只是取当前顺序里的上一行；所以排序是计算前的必要步骤。

In [80]:
# 先按公司和统计截止日期排序，确认面板数据的时间顺序。
panel_ts = analysis_df.sort_values(["证券代码", "统计截止日期"]).copy()  # 按公司和时间排序

panel_ts[["证券代码", "证券简称", "统计截止日期", "年份", "营业收入_亿元", "净利润_亿元"]].head(9)  # 查看排序后的面板

,证券代码,证券简称,统计截止日期,年份,营业收入_亿元,净利润_亿元
0,000002,万科A,2018-12-31,2018,2976.79,492.72
1,000002,万科A,2019-12-31,2019,3678.94,551.32
46,000002,万科A,2020-12-31,2020,4191.12,592.98
2,000004,国农科技,2018-12-31,2018,3.67,-0.22
3,000004,国农科技,2019-12-31,2019,1.08,-0.04
47,000004,国华网安,2020-12-31,2020,2.81,0.62
4,000006,深振业A,2018-12-31,2018,25.12,9.23
5,000006,深振业A,2019-12-31,2019,37.31,8.43
48,000006,深振业A,2020-12-31,2020,29.35,9.03


In [81]:
# 取出一家公司，先观察单个公司的时间序列。
company_ts = panel_ts.loc[panel_ts["证券简称"] == "万科A"].copy()  # 筛选万科A
company_ts[["证券代码", "证券简称", "统计截止日期", "年份", "营业收入_亿元", "净利润_亿元"]]  # 查看公司年度记录

,证券代码,证券简称,统计截止日期,年份,营业收入_亿元,净利润_亿元
0,000002,万科A,2018-12-31,2018,2976.79,492.72
1,000002,万科A,2019-12-31,2019,3678.94,551.32
46,000002,万科A,2020-12-31,2020,4191.12,592.98


**滞后、差分和增长率**

1. `shift(1)`：把数据向后移动一期，得到上一年的值。

2. `diff()`：计算本年和上一年的差。

3. `pct_change()`：计算本年相对上一年的增长率。

In [82]:
# 对单家公司计算上一年收入、收入增加额和收入增长率。
company_ts["上一年营业收入_亿元"] = company_ts["营业收入_亿元"].shift(1)  # 上一年收入
company_ts["收入增加额_亿元"] = company_ts["营业收入_亿元"].diff()  # 本年收入减上一年收入
company_ts["收入增长率"] = company_ts["营业收入_亿元"].pct_change()  # 收入增长率

company_ts[["证券简称", "年份", "营业收入_亿元", "上一年营业收入_亿元", "收入增加额_亿元", "收入增长率"]]  # 查看计算结果

,证券简称,年份,营业收入_亿元,上一年营业收入_亿元,收入增加额_亿元,收入增长率
0,万科A,2018,2976.79,NaN,NaN,NaN
1,万科A,2019,3678.94,2976.79,702.15,0.24
46,万科A,2020,4191.12,3678.94,512.18,0.14


**用增长率构造累计指数**

1. `(1 + 增长率)`：把增长率转换成增长倍数。

2. `cumprod()`：累计连乘。可以用增长率构造一个从 100 开始的收入指数。

3. `fillna(0)`：第一年的增长率缺失，这里把它当作 0，让指数从 100 开始。

In [83]:
# 用收入增长率构造收入指数。
company_ts["收入指数"] = 100 * (1 + company_ts["收入增长率"].fillna(0)).cumprod()  # 从 100 开始累计

company_ts[["证券简称", "年份", "营业收入_亿元", "收入增长率", "收入指数"]]  # 查看收入指数

,证券简称,年份,营业收入_亿元,收入增长率,收入指数
0,万科A,2018,2976.79,NaN,100.00
1,万科A,2019,3678.94,0.24,123.59
46,万科A,2020,4191.12,0.14,140.79


**在每家公司内部计算时间变化**

1. `groupby("证券代码")`：先按公司分组。

2. 分组后使用 `shift()`、`diff()`、`pct_change()`，表示在每家公司内部计算上一年、增加额和增长率。

3. 计算后要抽几家公司看一眼，确认不会把不同公司的年份接在一起。面板数据里直接对整列 `shift()` 很容易把上一家公司的最后一年错当成下一家公司的上一年。

In [84]:
# 在每家公司内部计算收入变化。
panel_ts["上一年营业收入_亿元"] = panel_ts.groupby("证券代码")["营业收入_亿元"].shift(1)  # 公司内部上一年收入
panel_ts["收入增加额_亿元"] = panel_ts.groupby("证券代码")["营业收入_亿元"].diff()  # 公司内部收入增加额
panel_ts["收入增长率"] = panel_ts.groupby("证券代码")["营业收入_亿元"].pct_change()  # 公司内部收入增长率

panel_ts[["证券代码", "证券简称", "年份", "营业收入_亿元", "上一年营业收入_亿元", "收入增加额_亿元", "收入增长率"]].head(12)  # 查看组内计算结果

,证券代码,证券简称,年份,营业收入_亿元,上一年营业收入_亿元,收入增加额_亿元,收入增长率
0,000002,万科A,2018,2976.79,NaN,NaN,NaN
1,000002,万科A,2019,3678.94,2976.79,702.15,0.24
46,000002,万科A,2020,4191.12,3678.94,512.18,0.14
2,000004,国农科技,2018,3.67,NaN,NaN,NaN
3,000004,国农科技,2019,1.08,3.67,-2.59,-0.71
47,000004,国华网安,2020,2.81,1.08,1.73,1.60
4,000006,深振业A,2018,25.12,NaN,NaN,NaN
5,000006,深振业A,2019,37.31,25.12,12.19,0.49
48,000006,深振业A,2020,29.35,37.31,-7.96,-0.21
6,000011,深物业A,2018,27.87,NaN,NaN,NaN


In [85]:
# 筛选 2020 年收入增长率较高的公司。
high_growth_2020 = panel_ts.loc[
    panel_ts["年份"] == 2020,
    ["证券代码", "证券简称", "行业名称", "营业收入_亿元", "上一年营业收入_亿元", "收入增长率"],
].sort_values("收入增长率", ascending=False)

high_growth_2020.head(10)  # 查看 2020 年增长率靠前的公司

,证券代码,证券简称,行业名称,营业收入_亿元,上一年营业收入_亿元,收入增长率
47,000004,国华网安,软件和信息技术服务业,2.81,1.08,1.60
63,000423,东阿阿胶,医药制造业,34.09,29.59,0.15
62,000403,派林生物,医药制造业,10.50,9.16,0.15
46,000002,万科A,房地产业,4191.12,3678.94,0.14
64,000550,江铃汽车,汽车制造业,330.96,291.74,0.13
60,000338,潍柴动力,汽车制造业,1974.91,1743.61,0.13
53,000021,深科技,计算机、通信和其他电子设备制造业,149.67,132.24,0.13
54,000030,富奥股份,汽车制造业,111.13,100.64,0.10
61,000400,许继电气,电气机械及器材制造业,111.91,101.56,0.10
57,000070,特发信息,电气机械及器材制造业,47.22,43.28,0.09


**日期索引和日期切片**

1. `set_index("日期列")`：把日期列设为索引。

2. 日期索引可以用字符串切片。例如年度数据可以用 `loc["2020"]` 取出 2020 年记录。

3. 本章数据是年度数据，不适合重点演示 `resample()`。遇到日度交易数据或月度宏观数据时，再用 `resample()` 改变频率。

In [86]:
# 把日期设为索引，演示按年份切片。
panel_by_date = panel_ts.set_index("统计截止日期").sort_index()  # 日期索引
panel_by_date.loc["2020", ["证券代码", "证券简称", "行业名称", "营业收入_亿元", "收入增长率"]].head()  # 取出 2020 年记录

,证券代码,证券简称,行业名称,营业收入_亿元,收入增长率
统计截止日期,,,,,
2020-12-31,000002,万科A,房地产业,4191.12,0.14
2020-12-31,000559,万向钱潮,汽车制造业,108.82,0.03
2020-12-31,000011,深物业A,房地产业,41.04,0.04
2020-12-31,000638,万方发展,软件和信息技术服务业,1.11,-0.04
2020-12-31,000555,神州信息,软件和信息技术服务业,106.86,0.05


阶段 8 小结：这一阶段带出了按时间排序、公司内部 `shift()`、`diff()`、`pct_change()`、用 `cumprod()` 构造累计指数、日期索引和日期切片。时间顺序计算完成后，先看单家公司，再看全部公司；这个顺序能减少很多面板数据中的错位问题。

**练习**：完成最后的阶段 8 练习。

## 练习

下面的练习继续使用本章两张表。建议每完成一步都看一眼结果，例如使用 `head()`、`tail()`、`shape` 或者简单统计。题目中的变量名是建议变量名，便于课堂核对。

### 阶段 1 练习：数据读取与初步查看

**练习 1.1**：查看经营表结构

读取 `company_annual_operations_clean.xlsx`，赋值给变量 `annual_ex1`。显示前 5 行和后 5 行，打印行列数、列名列表和各列数据类型。

**练习 1.2**：查看关键经营变量的分布

在 `annual_ex1` 中选择 `营业收入_亿元`、`净利润_亿元`、`总资产_亿元`、`资产负债率` 四列，赋值给变量 `key_business_ex1`。显示这四列的描述统计，并打印 2020 年样本行数。

### 阶段 2 练习：行列选择与条件筛选

**练习 2.1**：找出低负债且盈利的公司

暂时删除 `证券代码` 列，结果赋值给变量 `annual_no_code_ex2`。筛选 2020 年同时满足以下条件的公司：`净利润_亿元 > 0`、`资产负债率 < 0.6`。结果赋值给变量 `low_debt_profit_2020`，只保留 `股票简称`、`营业收入_亿元`、`净利润_亿元`、`资产负债率` 四列，并按 `资产负债率` 从低到高排序。打印结果行数，并显示前 5 行。

**练习 2.2**：选出收入和资产都较高的公司

在 2020 年公司中，筛选 `营业收入_亿元` 高于当年中位数、并且 `总资产_亿元` 高于当年中位数的公司，赋值给变量 `large_revenue_asset_2020`。显示 `股票简称`、`营业收入_亿元`、`总资产_亿元`、`资产负债率`，并查看前 5 行和后 5 行。

### 阶段 3 练习：变量生成与数据修改

**练习 3.1**：找出收入高但净利率较低的公司

在 `annual_no_code_ex2` 中新增 `净利率 = 净利润_亿元 / 营业收入_亿元`。筛选 2020 年 `营业收入_亿元` 高于当年中位数、同时 `净利率` 低于当年中位数的公司，赋值给变量 `high_revenue_low_margin`。这个结果代表“规模不小、利润率相对偏低”的公司。再为 2020 年公司按营业收入生成 `收入排名`。显示 `股票简称`、`营业收入_亿元`、`净利润_亿元`、`净利率`、`收入排名`，并查看前 5 行和后 5 行。

**练习 3.2**：给公司打上资产规模标签

用 `pd.cut()` 根据 `总资产_亿元` 生成 `资产规模`：`0-100` 为“小”，`100-1000` 为“中”，`1000` 以上为“大”。统计 2020 年不同 `资产规模` 的公司数量，赋值给变量 `size_counts_2020`。要求输出计数结果。

**练习 3.3**：按条件生成负债水平

在 `annual_no_code_ex2` 中新增 `负债水平`，先全部赋值为“正常”，再用 `.loc` 把 `资产负债率 >= 0.7` 的行改为“较高”。显示 `股票简称`、`年份`、`资产负债率`、`负债水平` 四列的前 8 行，并统计不同 `负债水平` 的数量。

### 阶段 4 练习：常见数据问题处理

**练习 4.1**：清洗证券代码和日期

重新读取 `financial_indicators_dirty.xlsx`，读取时把 `证券代码` 处理成字符串，赋值给变量 `dirty_indicators_ex4`。去掉 `证券代码` 和 `股票简称` 两端空格，把 `证券代码` 补齐为 6 位。把 `统计截止日期` 转成日期类型，并显示转换失败的记录。

**练习 4.2**：清理财务指标数值列

把 `加权平均净资产收益率`、`扣非净资产收益率`、`基本每股收益`、`非经常性损益_亿元`、`扣非净利润_亿元` 转成数值列。要求先处理百分号、逗号、`--`、`缺失` 和空字符串。显示转换后的数据类型，并展示至少一列清理前和清理后的对比结果。

**练习 4.3**：删除重复记录并生成盈利标签

按 `证券代码` 和 `统计截止日期` 删除重复记录，再删除这两个连接键缺失的记录，结果赋值给变量 `indicators_clean_ex4`。用 `扣非净利润_亿元 > 0` 生成 `扣非盈利`，统计 `扣非盈利` 的数量，并显示清理后每列缺失值数量。

### 阶段 5 练习：表的追加与合并

**练习 5.1**：纵向追加年度经营记录

重新读取 `company_annual_operations_clean.xlsx`，读取时把 `证券代码` 处理成 6 位字符串，并把 `股票简称` 重命名为 `证券简称`。生成 `年份`，把 2018-2019 年记录和 2020 年记录拆成两张表，再用 `concat()` 追加回一张表，赋值给变量 `annual_panel_ex5`。显示追加后的年度记录数。

**练习 5.2**：合并清洗后的指标表

把 `annual_panel_ex5` 和 `indicators_clean_ex4` 按 `证券代码`、`统计截止日期` 左连接，赋值给变量 `analysis_ex5`。检查合并前后行数是否一致，统计 `基本每股收益` 的缺失数量，并显示前 8 行。

**练习 5.3**：比较不同合并方式的行数

分别使用 `how="left"`、`how="inner"`、`how="outer"` 合并 `annual_panel_ex5` 和 `indicators_clean_ex4`，构造一张两列的小表：`how` 和 `行数`。显示这张表，并解释哪一种合并保留了经营表的所有记录。

### 阶段 6 练习：分组汇总与组内比较

**练习 6.1**：行业年度摘要表

把 `analysis_ex5` 复制为 `analysis_ex6`，并生成 `净利率`。按 `行业名称` 和 `年份` 分组，计算 `公司数`、`营业收入合计_亿元`、`平均净利率`、`平均资产负债率`，赋值给变量 `industry_year_summary`。显示前 12 行。

**练习 6.2**：每个行业找两家公司

在 2020 年数据中，按 `行业名称` 分组，找出每个行业 `营业收入_亿元` 最高的 2 家公司，赋值给变量 `top2_revenue_by_industry`。结果只保留 `行业名称`、`证券代码`、`证券简称`、`营业收入_亿元`、`净利率`，并按行业名称和营业收入排序。

**练习 6.3**：计算行业内部收入差距

自定义函数 `value_range(x)`，返回 `x.max() - x.min()`。按 `行业名称` 分组，计算 2020 年各行业营业收入的均值和收入差距，赋值给变量 `industry_gap_2020`。按收入差距从高到低排序。

**练习 6.4**：判断公司是否高于行业平均

在 2020 年数据中，用 `groupby()` 和 `transform()` 计算每家公司所在行业的平均营业收入，生成 `行业平均收入_亿元`。再生成 `高于行业平均`，表示该公司营业收入是否高于所在行业平均值。显示 `行业名称`、`证券简称`、`营业收入_亿元`、`行业平均收入_亿元`、`高于行业平均`，并查看前 10 行。

### 阶段 7 练习：数据重塑与结果表构造

**练习 7.1**：构造公司收入宽表

用 `pivot_table()` 把 `analysis_ex6` 整理为每家公司一行、年份为列、值为 `营业收入_亿元` 的宽表，赋值给变量 `revenue_wide_ex7`。计算 `收入增长率_2018_2020 = 2020 / 2018 - 1`，按增长率从高到低排序，显示前 10 行。

**练习 7.2**：构造公司综合摘要表

基于 `analysis_ex6` 构造 `company_profile_summary`，每家公司一行，至少包含：`证券代码`、`证券简称`、`行业名称`、`所属省份`、`2018` 年营业收入、`2020` 年营业收入、`收入增长率_2018_2020`、`2020` 年资产负债率。按 `收入增长率_2018_2020` 从高到低排序，显示前 10 行。

**练习 7.3**：核对公司摘要表

检查 `company_profile_summary` 的行数是否等于公司数量。再按 `行业名称` 统计公司数量，并显示统计结果。最后显示 `company_profile_summary` 的前 5 行和后 5 行。

### 阶段 8 练习：公司年度面板中的时间顺序

**练习 8.1**：计算单家公司收入变化

把 `analysis_ex6` 按 `证券代码` 和 `统计截止日期` 排序，赋值给变量 `panel_ex8`。选出任意一家公司，赋值给变量 `one_company_ex8`。计算 `上一年营业收入_亿元`、`收入增加额_亿元`、`收入增长率`，并显示该公司的全部年度记录。

**练习 8.2**：在公司内部计算收入增长率

在 `panel_ex8` 中按 `证券代码` 分组，计算每家公司内部的 `上一年营业收入_亿元` 和 `收入增长率`。筛选 2020 年收入增长率最高的 10 家公司，赋值给变量 `top_growth_2020_ex8`，显示 `证券代码`、`证券简称`、`行业名称`、`营业收入_亿元`、`上一年营业收入_亿元`、`收入增长率`。

**练习 8.3**：用日期索引筛选年度记录

把 `panel_ex8` 的 `统计截止日期` 设为索引并排序，赋值给变量 `panel_by_date_ex8`。用日期切片取出 2020 年记录，赋值给变量 `records_2020_ex8`。显示前 8 行，并统计 2020 年记录数。